pinot-core/src/main/java/com/linkedin/pinot/core/query/selection/SelectionOperatorService.java

```diff
@@ -392,450 +176,175 @@ public int compare(Serializable[] o1, Serializable[] o2) {
-    for (int i = 0; i < selectionColumnArray.length; ++i) {
-      String selectionColumn = selectionColumnArray[i];
-      if (!columns.contains(selectionColumn)) {
-        columns.add(selectionColumn);
-      }
-    }
-    final DataType[] dataTypes = new DataType[columns.size()];
-    for (int i = 0; i < dataTypes.length; ++i) {
-      DataSource ds = indexSegment.getDataSource(columns.get(i));
-      DataSourceMetadata m = ds.getDataSourceMetadata();
-      dataTypes[i] = m.getDataType();
-      if (!m.isSingleValue()) {
-        dataTypes[i] = DataType.valueOf(dataTypes[i] + "_ARRAY");
-      }
-    }
-    return new DataSchema(columns.toArray(new String[0]), dataTypes);
+  /**
+   * Get number of documents scanned.
+   *
+   * @return number of documents scanned.
+   */
+  public long getNumDocsScanned() {
+    return _numDocsScanned;
   }
 
-  public void iterateOnBlock(BlockDocIdIterator blockDocIdIterator, Block[] blocks)
-      throws Exception {
-    int docId = 0;
-    _rowDocIdComparator = getDocValBasedComparator(_sortSequence, _dataSchema, blocks);
-    if (_doOrdering) {
-      _rowDocIdSet = new PriorityQueue<Integer>(_maxRowSize, _rowDocIdComparator);
-    } else {
-      _rowDocIdSet = new ArrayList<Integer>(_maxRowSize);
+  /**
+   * Merge two partial results for selection query. (Server side)
+   *
+   * @param rowEventsSet1 partial results 1.
+   * @param rowEventsSet2 partial results 2.
+   * @return merged results.
+   */
+  public Collection<Serializable[]> merge(Collection<Serializable[]> rowEventsSet1,
+      Collection<Serializable[]> rowEventsSet2) {
+    if (rowEventsSet1 == null) {
+      return rowEventsSet2;
     }
-    while ((docId = blockDocIdIterator.next()) != Constants.EOF) {
-      _numDocsScanned++;
-      if (_rowDocIdSet.size() < _maxRowSize) {
-        _rowDocIdSet.add(docId);
-      } else {
-        if (!_doOrdering) {
-          break;
-        }
-        PriorityQueue<Integer> queue = (PriorityQueue<Integer>) _rowDocIdSet;
-        if (_doOrdering && (_rowDocIdComparator.compare(docId, queue.peek()) > 0)) {
-          queue.add(docId);
-          queue.poll();
-        }
-      }
+    if (rowEventsSet2 == null) {
+      return rowEventsSet1;
     }
-    mergeToRowEventsSet(blocks);
-  }
-
-  public Collection<Serializable[]> mergeToRowEventsSet(Block[] blocks)
-      throws Exception {
-    SelectionFetcher selectionFetcher = new SelectionFetcher(blocks, _dataSchema);
     if (_doOrdering) {
-      final PriorityQueue<Serializable[]> rowEventsPriorityQueue =
-          new PriorityQueue<Serializable[]>(_maxRowSize, _rowComparator);
-      PriorityQueue<Integer> queue = (PriorityQueue<Integer>) _rowDocIdSet;
-      while (!queue.isEmpty()) {
-        Serializable[] rowFromBlockValSets = selectionFetcher.getRow(queue.poll());
-        rowEventsPriorityQueue.add(rowFromBlockValSets);
+      PriorityQueue<Serializable[]> queue1 = (PriorityQueue<Serializable[]>) rowEventsSet1;
```

Comment:
Could you please rename the variables queue1/2 to reflect which is the source and which is the destination?